### ISSUE TITLE
Database Articles

### ISSUE THREAD

**Issue by benclmnt** [benclmnt/papers#3]

**Comment by benclmnt** [comment]
# [Uber's switch from PostgreSQL to MySQL (2016)](https://eng.uber.com/postgres-to-mysql-migration/)

**tl;dr:** Uber has a lot of secondary indexes. Some of Postgres MVCC design suffer from **write amplification** as all indexes needs to be updated on any tuple update + **replication amplification** as all update needs to be written to WAL, and WAL will need to be shipped to replicas.

- PostgreSQL (before version 10) uses physical replication via WAL shipping, whereas InnoDB supports statement-based replication & logical replication via WAL shipping. Logical replication causes less updates needed to be shipped, as other updates could be inferred from the replication updates itself.
- Since MySQL replication is decoupled from the physical representation,
    - it is harder to corrupt index records. The common problems are skipped statements / doubly-applied statements.
    - is easier to upgrade (can stream between different versions as long as it still uses the same replication format)
    - applying WAL updates do not block read queries on replicas (true MVCC replication)
- The table below shows how secondary indices are layout on disk and why Postgres suffers from write amplification.

| Postgres | InnoDB |
| --- | --- |
| ![postgres index records pointing to on-disk location](https://user-images.githubusercontent.com/49342399/146710918-84a2b70c-2b68-47ba-813b-bf4bd6387ae6.png) | ![innodb index records add a layer of indirection](https://user-images.githubusercontent.com/49342399/146710939-45b9e49b-d0c8-4827-8873-b2e1dfbcef2e.png) |

**Comment by benclmnt** [comment]
# [Figma's Vertical Partitioning](https://www.figma.com/blog/how-figma-scaled-to-multiple-databases/)

![image](https://github.com/benclmnt/papers/assets/49342399/bc456ac7-f39f-4c58-a172-dca19e49b999)

Figma has the largest single managed Postgres deployment. They decided against self-hosting and trying horizontal partitioning solutions. This article focused more on how they do their 2-step migrations: (1) PgBouncer and (2) data.
